From f47e4ef51a48db700542614c76674a8fa77f0e78 Mon Sep 17 00:00:00 2001
From: Yuan Juntao <juntao.yuan@intel.com>
Date: Thu, 15 Jan 2015 09:30:35 +0800
Subject: [PATCH] mmc: add sd card support for cht with dollar cove ti pmic

Add vmmc & vqmmc regulators for SD host controller with
dollar cove ti pmic.

Change-Id: I491e3a19b0b85288b6b7e86841068450c4e2e87f
Signed-off-by: Yuan Juntao <juntao.yuan@intel.com>
---
 .../device_libs/platform_sdio_regulator.c          | 37 ++++++++++++++++++++++
 drivers/mfd/intel_soc_pmic_dc_ti.c                 | 13 ++++++++
 2 files changed, 50 insertions(+)

diff --git a/arch/x86/platform/intel-mid/device_libs/platform_sdio_regulator.c b/arch/x86/platform/intel-mid/device_libs/platform_sdio_regulator.c
index 158d8c0..bb866c6 100644
--- a/arch/x86/platform/intel-mid/device_libs/platform_sdio_regulator.c
+++ b/arch/x86/platform/intel-mid/device_libs/platform_sdio_regulator.c
@@ -26,6 +26,7 @@
 #include <asm/cpu_device_id.h>
 #include <linux/regulator/intel_whiskey_cove_pmic.h>
 #include <linux/regulator/intel_dollar_cove_pmic.h>
+#include <linux/regulator/intel_dollar_cove_ti_pmic.h>
 
 struct acpi_ids { char *hid; char *uid; };
 
@@ -244,11 +245,47 @@ static void intel_setup_dollar_cove_sd_regulators(void)
 		sizeof(struct dcovex_regulator_info), DCOVEX_ID_LDO3 + 1);
 }
 
+/*************************************************************
+*
+* DCOVETI SD card related regulator
+*
+*************************************************************/
+static struct regulator_init_data dcoveti_vmmc_data;
+static struct regulator_init_data dcoveti_vqmmc_data;
+static struct dcoveti_regulator_info dcoveti_vmmc_info = {
+	.init_data = &dcoveti_vmmc_data,
+};
+
+static struct dcoveti_regulator_info dcoveti_vqmmc_info = {
+	.init_data = &dcoveti_vqmmc_data,
+};
+
+static void intel_setup_dollar_cove_ti_sd_regulators(void)
+{
+	memcpy((void *)&dcoveti_vmmc_data, (void *)&vmmc_data,
+			sizeof(struct regulator_init_data));
+	memcpy((void *)&dcoveti_vqmmc_data, (void *)&vqmmc_data,
+			sizeof(struct regulator_init_data));
+
+	/* set enable time for vqmmc regulator, stabilize power rail */
+	dcoveti_vqmmc_data.constraints.enable_time = 10000;
+
+	dcoveti_vmmc_data.constraints.name = "LDO_8";
+	dcoveti_vqmmc_data.constraints.name = "LDO_7";
+
+	/* register SD card regulator for dollar cove PMIC */
+	intel_soc_pmic_set_pdata("dcoveti_regulator", &dcoveti_vmmc_info,
+		sizeof(struct dcoveti_regulator_info), DCOVETI_ID_LDO8 + 1);
+	intel_soc_pmic_set_pdata("dcoveti_regulator", &dcoveti_vqmmc_info,
+		sizeof(struct dcoveti_regulator_info), DCOVETI_ID_LDO7 + 1);
+}
+
 static int __init sdio_regulator_init(void)
 {
 	intel_setup_crystal_cove_sd_regulators();
 	intel_setup_whiskey_cove_sd_regulators();
 	intel_setup_dollar_cove_sd_regulators();
+	intel_setup_dollar_cove_ti_sd_regulators();
 
 	return 0;
 }
diff --git a/drivers/mfd/intel_soc_pmic_dc_ti.c b/drivers/mfd/intel_soc_pmic_dc_ti.c
index 2a433f4..924b9ec 100644
--- a/drivers/mfd/intel_soc_pmic_dc_ti.c
+++ b/drivers/mfd/intel_soc_pmic_dc_ti.c
@@ -24,6 +24,7 @@
 #include <linux/version.h>
 
 #include "intel_soc_pmic_core.h"
+#include <linux/regulator/intel_dollar_cove_ti_pmic.h>
 
 #define IRQLVL1		0x01
 #define MIRQLVL1	0x02
@@ -119,6 +120,18 @@ static struct mfd_cell dollar_cove_ti_dev[] = {
 		.name = "dc_ti_region",
 		.id = 0,
 	},
+	{
+		.name = "dcoveti_regulator",
+		.id = DCOVETI_ID_LDO7 + 1,
+		.num_resources = 0,
+		.resources = NULL,
+	},
+	{
+		.name = "dcoveti_regulator",
+		.id = DCOVETI_ID_LDO8 + 1,
+		.num_resources = 0,
+		.resources = NULL,
+	},
 	{NULL, },
 };
 
-- 
1.9.1

